packages/tldraw/src/lib/shapes/shared/useEmojis.ts

```diff
@@ -0,0 +1,125 @@
+import { useState } from 'react'
+import { useDialogs } from '../../ui/hooks/useDialogsProvider'
+import EmojiDialog from './emojis'
+import { EmojiDialogSingleton } from './emojis/EmojiDialog'
+
+export function useEmojis(inputEl: HTMLTextAreaElement | null, onComplete: (text: string) => void) {
+	const { addDialog, removeDialog } = useDialogs()
+	const [emojiSearchText, setEmojiSearchText] = useState('')
+	const [isEmojiMenuOpen, setIsEmojiMenuOpen] = useState(false)
+
+	const closeMenu = () => {
+		setIsEmojiMenuOpen(false)
+		removeDialog('emoji')
+		setEmojiSearchText('')
+	}
+
+	const onEmojiSelect = (emoji: any) => {
+		if (!inputEl) return
+
+		const searchText = EmojiDialogSingleton?.component.refs.searchInput.current.value
+		inputEl.focus()
+		inputEl.setSelectionRange(
+			inputEl.selectionStart - searchText.length - 1,
+			inputEl.selectionStart
+		)
+		inputEl.setRangeText(emoji.native)
+		inputEl.setSelectionRange(inputEl.selectionStart + 1, inputEl.selectionStart + 1)
+		onComplete(inputEl.value)
+
+		closeMenu()
+	}
+
+	const onKeyDown = (
+		e: React.KeyboardEvent<HTMLTextAreaElement>,
+		coords: { top: number; left: number } | null
+	) => {
+		const emojiPicker = EmojiDialogSingleton?.component
+
+		switch (e.key) {
+			case ':': {
+				if (isEmojiMenuOpen) {
+					closeMenu()
+					return false
+				}
+
+				setEmojiSearchText('')
+				addDialog({
+					id: 'emoji',
+					component: EmojiDialog,
+					isCustomDialog: true,
+					dialogProps: {
+						onEmojiSelect,
+						onClickOutside: closeMenu,
+						top: coords?.top,
+						left: coords?.left,
+					},
+				})
+				setIsEmojiMenuOpen(true)
+
+				return true
+			}
+
+			case ' ':
+			// fall-through
+			case 'Escape': {
+				if (isEmojiMenuOpen) {
+					closeMenu()
+					return true
+				}
+
+				return false
+			}
+
+			case 'Enter':
+			case 'ArrowLeft':
+			case 'ArrowRight':
+			case 'ArrowUp':
+			case 'ArrowDown': {
+				if (isEmojiMenuOpen) {
+					emojiPicker.handleSearchKeyDown({
+						...e,
+						preventDefault: () => {
+							/* shim */
+						},
+						stopImmediatePropagation: () => {
+							/* shim */
+						},
+					})
+					e.preventDefault()
+					return true
+				}
+
+				return false
+			}
+
+			case 'Backspace':
```

Comment:
I don't love us effectively implementing manual text editing here... there's a lot missing (copy-paste, delete-key, etc).

In particular, I'd be curious about how this works with non-keyboard input methods (dictation, mobile pinyin keyboards, etc)

Reply:
An alternative here would be to mark the point at the start of the input and pull the latest range from there to the current caret? might have some issues with concurrent editing, but we don't officially support that anyway so

Reply:
agree to disagree here. i think you're overthinking it :)
i'm obviously borrowing the (exact) mechanics of Dropbox Paper's emoji picker which "steals" the keyboard events whilst being open.

i'm definitely aware of complexities around concurrent and IME (Input Method Editors) but that's out-of-scope for this feature set to be honest. those input methods are just separate beasts altogether :)